File: Python/第一章/06-context-window-demo/context_budget.py
"""演示如何在有限的 Context Window 内分配输入预算。"""

import math
from pprint import pprint
from typing import Any


MODEL_CONTEXT_LIMIT = 220
OUTPUT_RESERVE = 60
EXTERNAL_CONTEXT_RESERVE = 40

Message = dict[str, str]
Turn = list[Message]


def is_han_character(character: str) -> bool:
    """判断字符是否属于本示例需要处理的常见汉字范围。"""
    code_point = ord(character)
    return (
        0x3400 <= code_point <= 0x4DBF
        or 0x4E00 <= code_point <= 0x9FFF
        or 0xF900 <= code_point <= 0xFAFF
    )


def estimate_text_tokens(text: str) -> int:
    """粗略估算文本 Token 数量，不代表任何真实模型的 Tokenizer。"""
    chinese_count = sum(is_han_character(char) for char in text)
    other_count = len(text) - chinese_count
    return math.ceil(chinese_count * 0.7 + other_count / 4)


def estimate_message_tokens(message: Message) -> int:
    """估算一条消息正文以及 role、分隔符和模板的开销。"""
    return estimate_text_tokens(message["content"]) + 6


def estimate_messages_tokens(messages: list[Message]) -> int:
    """估算多条消息的 Token 总量。"""
    return sum(estimate_message_tokens(message) for message in messages)


def build_context(
    *,
    system_message: Message,
    summary_message: Message,
    history_turns: list[Turn],
    user_message: Message,
) -> dict[str, Any]:
    """保留必选消息，再从最近一轮开始装入完整历史轮次。"""
    input_budget = (
        MODEL_CONTEXT_LIMIT
        - OUTPUT_RESERVE
        - EXTERNAL_CONTEXT_RESERVE
    )
    required_messages = [system_message, summary_message, user_message]
    required_tokens = estimate_messages_tokens(required_messages)

    if required_tokens > input_budget:
        raise ValueError("必选内容已经超过输入预算，需要继续压缩摘要或当前输入。")

    remaining_budget = input_budget - required_tokens
    selected_turns: list[Turn] = []
    discarded_turns: list[Turn] = []

    for turn in reversed(history_turns):
        turn_tokens = estimate_messages_tokens(turn)
        if turn_tokens <= remaining_budget:
            selected_turns.insert(0, turn)
            remaining_budget -= turn_tokens
        else:
            discarded_turns.insert(0, turn)

    selected_messages = [
        message
        for turn in selected_turns
        for message in turn
    ]
    final_messages = [
        system_message,
        summary_message,
        *selected_messages,
        user_message,
    ]

    return {
        "messages": final_messages,
        "selectedTurns": selected_turns,
        "discardedTurns": discarded_turns,
        "inputBudget": input_budget,
        "usedInputTokens": estimate_messages_tokens(final_messages),
        "remainingInputTokens": remaining_budget,
    }


system_message = {
    "role": "system",
    "content": "你是星河零售公司的退款审核助手，只根据已提供的信息判断。",
}

summary_message = {
    "role": "system",
    "content": "历史摘要：用户正在处理订单 A1024，退款申请人为小明，退款金额为 3000 元。",
}

history_turns = [
    [
        {
            "role": "user",
            "content": "帮我介绍一下公司的退款流程，并给出每个环节的负责人。",
        },
        {
            "role": "assistant",
            "content": (
                "退款流程包括提交申请、规则校验、订单核对、人工审核和原路退款。"
                "不同退款类型会进入不同处理环节。"
            ),
        },
    ],
    [
        {
            "role": "user",
            "content": "订单 A1024 的退款金额是 3000 元。",
        },
        {
            "role": "assistant",
            "content": "已记录订单 A1024 的退款金额为 3000 元。",
        },
    ],
    [
        {
            "role": "user",
            "content": "退款金额超过 2000 元时，需要人工审核。",
        },
        {
            "role": "assistant",
            "content": "已记录该退款审核规则。",
        },
    ],
]

user_message = {
    "role": "user",
    "content": "订单 A1024 是否需要人工审核？只回答结论和依据。",
}


def run_demo() -> dict[str, Any]:
    """计算完整输入与裁剪后输入的预算数据。"""
    all_messages = [
        system_message,
        summary_message,
        *(message for turn in history_turns for message in turn),
        user_message,
    ]
    all_messages_tokens = estimate_messages_tokens(all_messages)
    result = build_context(
        system_message=system_message,
        summary_message=summary_message,
        history_turns=history_turns,
        user_message=user_message,
    )

    print("本次 Context Budget 分配结果：")
    pprint(
        {
            "模型窗口": MODEL_CONTEXT_LIMIT,
            "输出预留": OUTPUT_RESERVE,
            "外部上下文预留": EXTERNAL_CONTEXT_RESERVE,
            "本次输入预算": result["inputBudget"],
            "全部发送需要Token": all_messages_tokens,
            "裁剪后实际输入Token": result["usedInputTokens"],
            "剩余输入预算": result["remainingInputTokens"],
            "保留历史轮数": len(result["selectedTurns"]),
            "移出历史轮数": len(result["discardedTurns"]),
        },
        sort_dicts=False,
    )
    print("\n最终准备发送给模型的 messages：")
    pprint(result["messages"], sort_dicts=False)
    print("\n本次没有发送的旧历史：")
    pprint(result["discardedTurns"], sort_dicts=False)

    return {**result, "allMessagesTokens": all_messages_tokens}


if __name__ == "__main__":
    run_demo()

File: Node/第一章/06-context-window-demo/context-budget.js
// 为了让实验结果更明显，这里故意设置一个很小的窗口。
const MODEL_CONTEXT_LIMIT = 220

// Context Window 不只容纳输入，还需要为模型回答预留空间。
const OUTPUT_RESERVE = 60

// Agent 后续还可能加入 RAG 文档、工具描述和工具结果。
const EXTERNAL_CONTEXT_RESERVE = 40

/**
 * 粗略估算文字占用的 Token。
 * 这不是任何真实模型的 Tokenizer，只用于演示预算分配流程。
 */
function estimateTextTokens(text) {
	const characters = [...text]
	const chineseCharacterCount = characters.filter((character) =>
		/\p{Script=Han}/u.test(character)
	).length
	const otherCharacterCount = characters.length - chineseCharacterCount

	return Math.ceil(chineseCharacterCount * 0.7 + otherCharacterCount / 4)
}

function estimateMessageTokens(message) {
	// role、分隔符和消息模板同样会占用 Token。
	// 这里使用固定数值模拟这部分开销。
	return estimateTextTokens(message.content) + 6
}

function estimateMessagesTokens(messages) {
	return messages.reduce(
		(total, message) => total + estimateMessageTokens(message),
		0
	)
}

/**
 * 在输入预算内，组装本次真正发送给模型的 messages。
 */
function buildContext({
	systemMessage,
	summaryMessage,
	historyTurns,
	userMessage
}) {
	// 从完整窗口中，减去输出和外部上下文预留。
	const inputBudget =
		MODEL_CONTEXT_LIMIT - OUTPUT_RESERVE - EXTERNAL_CONTEXT_RESERVE

	// 这三部分是当前任务必须保留的内容。
	const requiredMessages = [systemMessage, summaryMessage, userMessage]
	const requiredTokens = estimateMessagesTokens(requiredMessages)

	if (requiredTokens > inputBudget) {
		throw new Error('必选内容已经超过输入预算，需要继续压缩摘要或当前输入。')
	}

	let remainingBudget = inputBudget - requiredTokens
	const selectedTurns = []
	const discardedTurns = []

	// 从最近一轮开始向前选择，每次保留完整的 user + assistant。
	for (let index = historyTurns.length - 1; index >= 0; index -= 1) {
		const turn = historyTurns[index]
		const turnTokens = estimateMessagesTokens(turn)

		if (turnTokens <= remainingBudget) {
			selectedTurns.unshift(turn)
			remainingBudget -= turnTokens
		} else {
			discardedTurns.unshift(turn)
		}
	}

	const messages = [
		systemMessage,
		summaryMessage,
		...selectedTurns.flat(),
		userMessage
	]

	return {
		messages,
		selectedTurns,
		discardedTurns,
		inputBudget,
		usedInputTokens: estimateMessagesTokens(messages),
		remainingInputTokens: remainingBudget
	}
}

const systemMessage = {
	role: 'system',
	content: '你是星河零售公司的退款审核助手，只根据已提供的信息判断。'
}

// 历史摘要保留旧对话中仍然重要的事实。
// 这里直接准备摘要，后续 Memory 章节会继续学习如何生成和保存它。
const summaryMessage = {
	role: 'system',
	content:
		'历史摘要：用户正在处理订单 A1024，退款申请人为小明，退款金额为 3000 元。'
}

// 按完整对话轮次保存历史，避免只留下问题或者回答。
const historyTurns = [
	[
		{
			role: 'user',
			content: '帮我介绍一下公司的退款流程，并给出每个环节的负责人。'
		},
		{
			role: 'assistant',
			content:
				'退款流程包括提交申请、规则校验、订单核对、人工审核和原路退款。不同退款类型会进入不同处理环节。'
		}
	],
	[
		{
			role: 'user',
			content: '订单 A1024 的退款金额是 3000 元。'
		},
		{
			role: 'assistant',
			content: '已记录订单 A1024 的退款金额为 3000 元。'
		}
	],
	[
		{
			role: 'user',
			content: '退款金额超过 2000 元时，需要人工审核。'
		},
		{
			role: 'assistant',
			content: '已记录该退款审核规则。'
		}
	]
]

const userMessage = {
	role: 'user',
	content: '订单 A1024 是否需要人工审核？只回答结论和依据。'
}

// 先计算：如果什么都不删除，全部发送需要多少 Token。
const allMessages = [
	systemMessage,
	summaryMessage,
	...historyTurns.flat(),
	userMessage
]
const allMessagesTokens = estimateMessagesTokens(allMessages)

// 再根据预算，组装本次真正需要发送的内容。
const result = buildContext({
	systemMessage,
	summaryMessage,
	historyTurns,
	userMessage
})

console.log('本次 Context Budget 分配结果：')
console.table({
	模型窗口: MODEL_CONTEXT_LIMIT,
	输出预留: OUTPUT_RESERVE,
	外部上下文预留: EXTERNAL_CONTEXT_RESERVE,
	本次输入预算: result.inputBudget,
	全部发送需要Token: allMessagesTokens,
	裁剪后实际输入Token: result.usedInputTokens,
	剩余输入预算: result.remainingInputTokens,
	保留历史轮数: result.selectedTurns.length,
	移出历史轮数: result.discardedTurns.length
})

console.log('\n最终准备发送给模型的 messages：')
console.dir(result.messages, { depth: null })

console.log('\n本次没有发送的旧历史：')
console.dir(result.discardedTurns, { depth: null })
